File: test-suite/src/test/java/io/micronaut/kafka/docs/seek/aware/ProductListener.java
package io.micronaut.kafka.docs.seek.aware;

import io.micronaut.configuration.kafka.ConsumerSeekAware;
import io.micronaut.configuration.kafka.annotation.*;
import io.micronaut.configuration.kafka.seek.*;
import io.micronaut.context.annotation.Requires;
import io.micronaut.kafka.docs.Product;
import org.apache.kafka.common.TopicPartition;
import java.util.*;

@KafkaListener
@Requires(property = "spec.name", value = "ConsumerSeekAwareTest")
public class ProductListener implements ConsumerSeekAware { // <1>

    List<Product> processed = new ArrayList<>();

    public ProductListener(ProductListenerConfiguration config) {
        // ...
    }

    @Topic("wonderful-products")
    void receive(Product product) {
        processed.add(product);
    }

    @Override
    public void onPartitionsRevoked(Collection<TopicPartition> partitions) { // <2>
        // save offsets here
    }

    @Override
    public void onPartitionsAssigned(Collection<TopicPartition> partitions, KafkaSeeker seeker) { // <3>
        // seek to offset here
        partitions.stream().map(tp -> KafkaSeekOperation.seek(tp, 1)).forEach(seeker::perform);
    }
}

File: test-suite-python/src/test/python/micronaut/kafka/docs/seek/aware/ProductListener.py
from micronaut.configuration.kafka import ConsumerSeekAware
from micronaut.configuration.kafka.annotation import KafkaListener, Topic
from micronaut.configuration.kafka.seek import KafkaSeeker, KafkaSeekOperation
from micronaut.context.annotation import Requires
from micronaut.kafka.docs.Product import Product
from org.apache.kafka.common import TopicPartition

from .ProductListenerConfiguration import ProductListenerConfiguration


@KafkaListener
@Requires(property="spec.name", value="ConsumerSeekAwareTest")
class ProductListener(ConsumerSeekAware):  # <1>

    def __init__(self, config: ProductListenerConfiguration):
        self.processed: list[Product] = []

    @Topic("wonderful-products")
    def receive(self, product: Product) -> None:
        self.processed.append(product)

    def onPartitionsRevoked(self, partitions: list[TopicPartition]) -> None:  # <2>
        # save offsets here
        pass

    def onPartitionsAssigned(self, partitions: list[TopicPartition], seeker: KafkaSeeker) -> None:  # <3>
        # seek to offset here
        for tp in partitions:
            seeker.perform(KafkaSeekOperation.seek(tp, 1))
